modules/consensus/changelog.go

```diff
@@ -0,0 +1,162 @@
+package consensus
+
+// changelog.go implements a persistent changelog in the consenus database
+// tracking all of the atomic changes to the consensus set. The primary use of
+// the changelog is for subscribers that have persistence - instead of
+// subscribing from the very beginning and receiving all changes from genesis
+// each time the daemon starts up, the subscribers can start from the most
+// recent change that they are familiar with.
+//
+// The changelog is set up as a singley linked list where each change points
+// forward to the next change. In bolt, the key is a hash of the changeEntry
+// and the value is a struct containing the changeEntry and the key of the next
+// changeEntry. The empty hash key leads to the 'changeTail', which contains
+// the id of the most recent changeEntry.
+//
+// Initialization only needs to worry about creating the blank change entry,
+// the genesis block will call 'append' later on during initialization.
+
+import (
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+
+	"github.com/NebulousLabs/bolt"
+)
+
+var (
+	// ChangeLog contains a list of atomic changes that have happened to the
+	// consensus set so that subscribers can subscribe from the most recent
+	// change they have seen.
+	ChangeLog = []byte("ChangeLog")
+
+	// ChangeLogTailID is a key that points to the id of the current changelog
+	// tail.
+	ChangeLogTailID = []byte("ChangeLogTailID")
+)
+
+type (
+	// changeEntry records a single atomic change to the consensus set.
+	changeEntry struct {
+		RevertedBlocks []types.BlockID
+		AppliedBlocks  []types.BlockID
+	}
+
+	// changeNode contains a change entry and a pointer to the next change
+	// entry, and is the object that gets stored in the database.
+	changeNode struct {
+		Entry changeEntry
+		Next  modules.ConsensusChangeID
+	}
```

Comment:
they could, but then you'd be relying on indices, which I felt was not safe enough, especially if people are going to be copying over their consensus folder from other directories or whatever sillyness they might get into. Instead, they get an ID which points to an exact changeEntry. If the changeEntry exists, then they are guaranteed to be reading something that's sensical. If it's just an index, there's not the same certainty that the element at that index is the element you are expecting.